egui_demo_lib/src/demo/autocomplete.rs

```diff
@@ -0,0 +1,266 @@
+        Grid::new("autocomplete_grid")
+            .num_columns(2)
+            .show(ui, |ui| {
+                ui.label("Source");
+                ui.text_edit_multiline(&mut self.autocomplete_source);
+                ui.end_row();
+                ui.label("Input");
+                let te_id = ui.make_persistent_id("text_edit_ac");
+                let up_pressed = ui
+                    .input_mut()
+                    .consume_key(Modifiers::default(), Key::ArrowUp);
+                let down_pressed = ui
+                    .input_mut()
+                    .consume_key(Modifiers::default(), Key::ArrowDown);
+                let te = TextEdit::singleline(&mut self.input)
+                    .lock_focus(true)
+                    .id(te_id);
+                if self.cursor_to_end {
+                    text_edit_cursor_set_to_end(ui, te_id);
+                }
+                let re = ui.add(te);
+                self.ac_state.input_changed = re.changed();
+                if self.first_frame {
+                    re.request_focus();
+                    self.first_frame = false;
+                }
+                ui.end_row();
+                let candidates = self
+                    .autocomplete_source
+                    .split_whitespace()
+                    .collect::<Vec<_>>();
+                let msg = autocomplete_popup_below(
+                    &mut self.input,
+                    &mut self.ac_state,
+                    &candidates,
+                    ui,
+                    &re,
+                    up_pressed,
+                    down_pressed,
+                );
+                if msg.applied {
+                    self.cursor_to_end = true;
+                } else {
+                    self.cursor_to_end = false;
+                }
+                if ui.input().key_pressed(Key::Enter) {
+                    self.accepted_string = self.input.take();
+                    re.request_focus();
+                }
+                if msg.stole_focus {
+                    re.request_focus();
+                }
+                ui.label("Accepted");
+                ui.add(Label::new(&self.accepted_string).wrap(true));
+                ui.end_row();
+            });
+        ui.separator();
+        ui.vertical_centered(|ui| {
+            ui.label("Up/Down: navigate list");
+            ui.label("Tab: autocomplete");
+            ui.label("Enter: autocomplete/accept");
+        });
+    }
+}
+
+fn text_edit_cursor_set_to_end(ui: &mut egui::Ui, te_id: egui::Id) {
+    let mut state = TextEdit::load_state(ui.ctx(), te_id).unwrap();
+    state.set_cursor_range(Some(CursorRange::one(Cursor {
+        ccursor: CCursor {
+            index: 0,
+            prefer_next_row: false,
+        },
+        rcursor: RCursor { row: 0, column: 0 },
+        pcursor: PCursor {
+            paragraph: 0,
+            offset: 10000,
+            prefer_next_row: false,
+        },
+    })));
```

Comment:
If you want to set the cursor to the end, this is the way to do it:

``` rust
let ccursor = egui::text::CCursor::new(text.chars().count());
state.set_ccursor_range(Some(egui::text::CCursorRange::one(ccursor)));
```

Reply:
Well, I think it's fine to just set the position to the end for now.